### PR TITLE
fix: properly detect, specify, and handle connecting forked network to non-fork  [APE-1393]

### PR DESCRIPTION
**Pull request by antazoey**
### What I did

* add method to get metadata, which included a fork object that is None when not a fork
* Raises error when connecting fork to non fork like we use to, but use new way that is not as bugged
* Allow setting the host per forked network so you can do that if you want
* Removed a deprecated part from the README
* Removes unneeded code
* Updates lint dependencies

fixes: https://github.com/ApeWorX/ape/issues/1467

### Checklist

- [ ] Passes all linting checks (pre-commit and CI jobs)
- [ ] New test cases have been added and are passing
- [ ] Documentation has been updated
- [ ] PR title follows [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) standard (will be automatically included in the changelog)

### CODE REVIEW

**Comment by antazoey** [comment]
Partially addresses: https://github.com/ApeWorX/ape/issues/1467

File: ape_hardhat/provider.py [hunk 1]
```diff
@@ -918,28 +967,15 @@ def _upstream_provider(self) -> ProviderAPI:
...
-        self._upstream_provider.disconnect()
-
-        if self.get_block(0).hash != upstream_genesis_block_hash:
-            logger.warning(
-                "Upstream network has mismatching genesis block. "
-                "This could be an issue with hardhat."
+        if not self.metadata.forked_network:
+            # This will fail when trying to connect hardhat-fork to
+            # a non-forked network.
+            raise HardhatProviderError(
```
Comment (antazoey) [inline ape_hardhat/provider.py:973]: so we had an issue with the genesis block on certain networks, which is why we had to switch to logging as an error instead, with hopes to someday having a better way. Now, hardhat has a metadata RPC that we can use to deduce whether the network is a fork, so we can put back this error.

See this comment for historical reference: https://github.com/ApeWorX/ape-hardhat/pull/32#issuecomment-1073026712

File: ape_hardhat/provider.py [hunk 2]
```diff
@@ -918,28 +967,15 @@ def _upstream_provider(self) -> ProviderAPI:
...
+                "Network is no a fork. "
+                "Hardhat is likely already running on the local network. "
+                "Try using config:\n\n(ape-config.yaml)\n```\nhardhat:\n  "
+                "host: auto\n```\n\nso that multiple processes can automatically "
```
Comment (antazoey) [inline ape_hardhat/provider.py:977]: this message should prevent the user from being misled so much, and even has a suggestion for how to fix.
this is the part that fixes https://github.com/ApeWorX/ape/issues/1467

File: tests/test_provider.py [hunk 3]
```diff
@@ -176,26 +175,6 @@ def test_contract_revert_custom_exception(owner, get_contract_type, accounts):
     assert err.value.inputs == {"addr": accounts[7].address, "counter": 123}
 
 
-def test_contract_dev_message(owner, get_contract_type):
```
Comment (antazoey) [inline tests/test_provider.py:179]: unfortunately, we can longer run this test without installing `ape-vyper`, maybe we will find a way to have a less integration-y test

**Review by fubuloubu — APPROVED** [review]
Looks solid

**Review by dtdang — APPROVED** [review]

**Pull request merged**